File: ios/Tehilim/Core/Services/MemorialCalculator.swift
import Foundation

/// Calcule la prochaine azcara (anniversaire hébraïque du décès) à partir
/// d'une date civile de décès, en appliquant les règles traditionnelles :
///
///   - Adar / Adar II : dans une année embolismique, l'azcara d'un décès
///     survenu en Adar (année commune) est observée en Adar II.
///   - Heshvan 30 / Kislev 30 : si l'année cible n'a pas 30 jours dans
///     ce mois, l'azcara passe au 1er du mois suivant.
///   - Adar I 30 : si l'année cible n'est pas embolismique, l'azcara est
///     observée le 30 Shevat (tradition la plus répandue).
///
/// La logique est volontairement isolée de l'UI pour être testable. Le
/// pendant Android (`MemorialCalculator.kt`) reproduit exactement le même
/// algorithme avec ICU `HebrewCalendar`.
///
/// V1.10.7 — feature Commémoration.
enum MemorialCalculator {

    /// API publique : civile → civile. Retourne la date civile de la
    /// prochaine azcara strictement postérieure à `now`.
    static func nextYahrzeit(deathCivil: Date, now: Date = Date()) -> Date? {
        let death = hebrewYMD(from: deathCivil)
        let today = hebrewYMD(from: now)
        let sourceLeap = isLeap(year: death.year)

        // Borne de sécurité : en pratique 1 itération, max 2 si l'azcara
        // de l'année courante est déjà passée.
        for offset in 0..<3 {
            let targetYear = today.year + offset
            let targetLeap = isLeap(year: targetYear)
            let tm = adjustedMonth(
                sourceMonth: death.month,
                sourceLeap: sourceLeap,
                targetLeap: targetLeap
            )
            let (fm, fd) = adjustedDay(
                sourceDay: death.day,
                sourceMonth: death.month,
                sourceLeap: sourceLeap,
                targetMonth: tm,
                targetYear: targetYear
            )
            if let candidate = civilDate(year: targetYear, month: fm, day: fd),
               candidate > now {
                return candidate
            }
        }
        return nil
    }

    /// Convertit une date civile en composantes hébraïques.
    static func hebrewYMD(from date: Date) -> HebrewYMD {
        let cal = Calendar(identifier: .hebrew)
        let c = cal.dateComponents([.year, .month, .day], from: date)
        return HebrewYMD(year: c.year ?? 0, month: c.month ?? 0, day: c.day ?? 0)
    }

    /// Convertit une date hébraïque (Y/M/D Apple-indexée) en date civile
    /// (midi local pour éviter les ambiguïtés de bordure de jour).
    static func civilDate(year: Int, month: Int, day: Int) -> Date? {
        let cal = Calendar(identifier: .hebrew)
        var dc = DateComponents()
        dc.year = year
        dc.month = month
        dc.day = day
        dc.hour = 12
        return cal.date(from: dc)
    }

    /// Cycle de Méton : 7 années embolismiques sur 19. La formule
    /// `(7y + 1) mod 19 < 7` est la règle classique.
    static func isLeap(year: Int) -> Bool {
        ((7 * year + 1) % 19) < 7
    }

    /// Nombre de jours dans `month` pour `year`. Utilise NSCalendar pour
    /// rester aligné sur l'implémentation Apple (qui gère deficient /
    /// regular / complete year — l'utilisateur n'a pas à connaître).
    static func daysInMonth(month: Int, year: Int) -> Int {
        let cal = Calendar(identifier: .hebrew)
        var dc = DateComponents()
        dc.year = year
        dc.month = month
        dc.day = 1
        guard let date = cal.date(from: dc),
              let range = cal.range(of: .day, in: .month, for: date) else {
            return 29
        }
        return range.count
    }

    // MARK: - Règles métier (internes, exposées en `internal` pour les tests)

    /// Adaptation du mois entre année source et année cible.
    ///
    /// **Indexation Apple Hebrew (= ICU 1-indexé)** — STABLE quel que soit
    /// le statut leap de l'année :
    /// - 1 = Tishri, 2 = Heshvan, 3 = Kislev, 4 = Tevet, 5 = Shevat
    /// - **6 = Adar I** — n'existe QUE dans les années embolismiques
    /// - **7 = Adar** — en année commune = Adar unique ; en année
    ///                 embolismique = Adar II
    /// - 8 = Nisan, 9 = Iyar, 10 = Sivan, 11 = Tammuz, 12 = Av, 13 = Elul
    ///
    /// Conséquence : en année commune, le mois 6 est *skipped* (les mois
    /// retournés sont {1,2,3,4,5,7,8,9,10,11,12}). Pas de re-numérotation.
    /// Donc la seule transformation nécessaire est pour Adar I → Adar
    /// quand on passe d'une année leap à une non-leap.
    ///
    /// V1.10.7 — fix bug remonté en test : ancienne logique faisait des
    /// décalages ±1 pour les mois post-Adar (en supposant à tort une
    /// indexation contiguë), résultant en off-by-one (Sivan → Tammuz, etc.).
    static func adjustedMonth(
        sourceMonth: Int,
        sourceLeap: Bool,
        targetLeap: Bool
    ) -> Int {
        // Source Adar I (mois 6, leap-only) + cible commune → Adar (7).
        if sourceLeap && sourceMonth == 6 && !targetLeap {
            return 7
        }
        // Tous les autres cas (incluant Adar source → Adar II target leap
        // = même indice 7) : indice stable.
        return sourceMonth
    }

    /// Adaptation du jour quand le mois cible est plus court (29 vs 30).
    static func adjustedDay(
        sourceDay: Int,
        sourceMonth: Int,
        sourceLeap: Bool,
        targetMonth: Int,
        targetYear: Int
    ) -> (month: Int, day: Int) {
        let daysAvail = daysInMonth(month: targetMonth, year: targetYear)
        if sourceDay <= daysAvail {
            return (targetMonth, sourceDay)
        }
        // sourceDay == 30, targetMonth a 29 jours.
        // Cas spécial : 30 Adar I (source leap, sourceMonth == 6) →
        // 30 Shevat (mois 5) en année cible non-leap (Adar I n'existe pas).
        if sourceLeap && sourceMonth == 6 && sourceDay == 30 {
            return (5, 30)
        }
        // Règle standard : 30 Heshvan (2) → 1 Kislev (3),
        //                  30 Kislev (3)  → 1 Tevet (4).
        return (targetMonth + 1, 1)
    }
}

File: android/app/src/main/java/com/david/tehilim/core/service/MemorialCalculator.kt
package com.david.tehilim.core.service

import android.icu.util.Calendar as ICUCalendar
import android.icu.util.HebrewCalendar
import com.david.tehilim.core.model.HebrewYMD
import java.util.Date

/**
 * Calcule la prochaine azcara (anniversaire hébraïque du décès) à partir
 * d'une date civile, en appliquant les règles traditionnelles. Mirror
 * exact de `MemorialCalculator.swift` (iOS) — même algorithme.
 *
 * Indexation mois ICU (différente d'Apple Hebrew !) :
 * - 0 = TISHRI
 * - 1 = HESHVAN
 * - 2 = KISLEV
 * - 3 = TEVET
 * - 4 = SHEVAT
 * - 5 = ADAR_1  (n'existe QUE dans les années embolismiques)
 * - 6 = ADAR    (= Adar unique en année commune, = Adar II en embolismique)
 * - 7 = NISAN
 * - ...
 * - 12 = ELUL
 *
 * Règles métier :
 * - Adar (année commune) → Adar II en année cible embolismique
 * - Adar I leap → Adar unique en année cible commune
 * - Adar II leap → Adar unique en année cible commune
 * - 30 Heshvan / 30 Kislev avec mois cible 29 jours → 1er du mois suivant
 * - 30 Adar I leap → 30 Shevat en année cible commune
 *
 * V1.4 — feature Commémoration.
 */
object MemorialCalculator {

    /** API publique : civile → civile. */
    fun nextYahrzeit(deathCivil: Date, now: Date = Date()): Date? {
        val death = hebrewYMD(deathCivil)
        val today = hebrewYMD(now)
        val sourceLeap = isLeap(death.year)

        // Borne de sécurité : 1 itération en pratique, max 2 si déjà passé.
        for (offset in 0..2) {
            val targetYear = today.year + offset
            val targetLeap = isLeap(targetYear)
            val tm = adjustedMonth(death.month, sourceLeap, targetLeap)
            val (fm, fd) = adjustedDay(
                sourceDay = death.day,
                sourceMonth = death.month,
                sourceLeap = sourceLeap,
                targetMonth = tm,
                targetYear = targetYear
            )
            val candidate = civilDate(targetYear, fm, fd) ?: continue
            if (candidate.after(now)) return candidate
        }
        return null
    }

    /** Convertit une date civile en composantes hébraïques (ICU-indexées). */
    fun hebrewYMD(date: Date): HebrewYMD {
        val cal = HebrewCalendar()
        cal.time = date
        return HebrewYMD(
            year = cal.get(ICUCalendar.YEAR),
            month = cal.get(ICUCalendar.MONTH),
            day = cal.get(ICUCalendar.DAY_OF_MONTH)
        )
    }

    /** Convertit une date hébraïque en date civile (midi local). */
    fun civilDate(year: Int, month: Int, day: Int): Date? {
        return try {
            val cal = HebrewCalendar()
            cal.clear()
            cal.set(ICUCalendar.YEAR, year)
            cal.set(ICUCalendar.MONTH, month)
            cal.set(ICUCalendar.DAY_OF_MONTH, day)
            cal.set(ICUCalendar.HOUR_OF_DAY, 12)
            cal.set(ICUCalendar.MINUTE, 0)
            cal.set(ICUCalendar.SECOND, 0)
            cal.time
        } catch (e: Throwable) {
            null
        }
    }

    /** Cycle de Méton : 7 années embolismiques sur 19. */
    fun isLeap(year: Int): Boolean = ((7 * year + 1) % 19) < 7

    /** Nombre de jours dans `month` pour `year` (ICU-aware). */
    fun daysInMonth(month: Int, year: Int): Int {
        val cal = HebrewCalendar()
        cal.clear()
        cal.set(ICUCalendar.YEAR, year)
        cal.set(ICUCalendar.MONTH, month)
        cal.set(ICUCalendar.DAY_OF_MONTH, 1)
        return cal.getActualMaximum(ICUCalendar.DAY_OF_MONTH)
    }

    // Règles métier (exposées pour les tests)

    /**
     * Adaptation du mois entre année source et année cible. Indexation ICU.
     *
     * - sourceMonth == 5 (ADAR_1) : ne peut arriver qu'en source leap.
     *   Cible leap → ADAR_1 (5). Cible non-leap → ADAR (6, unique).
     * - sourceMonth == 6 (ADAR) : sémantique dépend de sourceLeap.
     *   - sourceLeap = true → c'est Adar II. Cible leap → ADAR (6) = Adar II.
     *     Cible non-leap → ADAR (6) = Adar unique.
     *   - sourceLeap = false → c'est Adar unique. Cible leap → ADAR (6) = Adar II
     *     (tradition Ashkenazi). Cible non-leap → ADAR (6) = inchangé.
     * - autres mois : inchangés (ICU n'a pas de décalage d'indexation, contraire-
     *   ment à Apple Hebrew).
     */
    fun adjustedMonth(sourceMonth: Int, sourceLeap: Boolean, targetLeap: Boolean): Int {
        return when {
            sourceMonth == 5 && sourceLeap -> if (targetLeap) 5 else 6
            // Tous les autres cas Adar (et tous les autres mois) → 6 ou inchangé.
            // ICU est "stable" : ADAR (6) reste 6 que l'année soit leap ou non.
            // Donc rien à transformer pour les mois post-Adar (7..12).
            else -> sourceMonth
        }
    }

    /** Adaptation du jour quand le mois cible est plus court (29 vs 30). */
    fun adjustedDay(
        sourceDay: Int,
        sourceMonth: Int,
        sourceLeap: Boolean,
        targetMonth: Int,
        targetYear: Int
    ): Pair<Int, Int> {
        val daysAvail = daysInMonth(targetMonth, targetYear)
        if (sourceDay <= daysAvail) return targetMonth to sourceDay

        // sourceDay == 30 mais targetMonth a 29 jours.
        // Cas spécial : 30 Adar I leap (ICU mois 5) + cible non-leap →
        // 30 Shevat (ICU mois 4) — tradition.
        if (sourceLeap && sourceMonth == 5 && sourceDay == 30) {
            return 4 to 30
        }
        // 30 Heshvan (1) → 1 Kislev (2). 30 Kislev (2) → 1 Tevet (3).
        return (targetMonth + 1) to 1
    }
}
